src/Adapter/CatalogPriceRule/QueryHandler/GetCatalogPriceRuleListHandler.php

```diff
@@ -0,0 +1,84 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\CatalogPriceRule\QueryHandler;
+
+use PrestaShop\Decimal\DecimalNumber;
+use PrestaShop\PrestaShop\Adapter\CatalogPriceRule\Repository\CatalogPriceRuleRepository;
+use PrestaShop\PrestaShop\Core\Domain\CatalogPriceRule\Query\GetCatalogPriceRuleList;
+use PrestaShop\PrestaShop\Core\Domain\CatalogPriceRule\QueryHandler\GetCatalogPriceRuleListHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\CatalogPriceRule\QueryResult\CatalogPriceRuleForListing;
+use PrestaShop\PrestaShop\Core\Domain\CatalogPriceRule\QueryResult\CatalogPriceRuleList;
+use PrestaShop\PrestaShop\Core\Domain\Product\SpecificPrice\Query\GetSpecificPriceForEditing;
+use PrestaShop\PrestaShop\Core\Util\DateTime\DateTime as DateTimeUtil;
+
+/**
+ * Handles @see GetSpecificPriceForEditing using legacy object model
+ */
+class GetCatalogPriceRuleListHandler implements GetCatalogPriceRuleListHandlerInterface
+{
+    /**
+     * @var CatalogPriceRuleRepository
+     */
+    private $catalogPriceRuleRepository;
+
+    /**
+     * @param CatalogPriceRuleRepository $catalogPriceRuleRepository
+     */
+    public function __construct(
+        CatalogPriceRuleRepository $catalogPriceRuleRepository
+    ) {
+        $this->catalogPriceRuleRepository = $catalogPriceRuleRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(GetCatalogPriceRuleList $query): CatalogPriceRuleList
+    {
+        $catalogPriceRules = $this->catalogPriceRuleRepository->getAll($query->getLangId());
+        $return = [];
+        foreach ($catalogPriceRules as $catalogPriceRule) {
+            $return[] = new CatalogPriceRuleForListing(
+                (int) $catalogPriceRule['id_specific_price_rule'],
+                $catalogPriceRule['specific_price_rule_name'],
+                $catalogPriceRule['shop_name'],
+                $catalogPriceRule['symbol'],
+                $catalogPriceRule['lang_name'],
+                $catalogPriceRule['group_name'],
+                (int) $catalogPriceRule['from_quantity'],
+                $catalogPriceRule['reduction_type'],
+                new DecimalNumber($catalogPriceRule['reduction']),
+                DateTimeUtil::buildNullableDateTime($catalogPriceRule['from']),
+                DateTimeUtil::buildNullableDateTime($catalogPriceRule['to'])
+            );
+        }
+
+        return new CatalogPriceRuleList($return, count($return));
```

Comment:
We don't have pagination right now, so the count is not really useful currently But it could be for shops that have many cart rules to avoid getting hundreds of cart rules at a time.

I'll let @MatShir decide if the pagination here is relevant, but in any case either the query handles it completely or not at all, this intermediate state will need to be fixed regardless of the decision.

Reply:
But even if we decide on a simple UX for no I still think the query should handle the pagination arguments, it will avoid future BC breaks because we will probably need them at some point And we can always force it to be limitless in the controller for now

Reply:
The specific price pagination has been integrated in this PR https://github.com/PrestaShop/PrestaShop/pull/28749 So you can integrate it in this PR as well now, and the pagination should also be behat tested